### ISSUE TITLE
Add more Prometheus metrics

### ISSUE THREAD

**Issue by yuvipanda** [jupyter/notebook#3682]
Now that #3490 has been merged, we should add more prometheus metrics to the notebook server!

Some ideas for metrics to add...

1. Number of kernels running (labeled by type)
2. Number of sessions open (not sure if this is useful?)
3. Number of terminals open
4. Mirror of the [activity tracking](https://github.com/jupyter/notebook/issues/1241) metrics
5. Kernel start / stop latency metrics
6. 0mq metrics

Am sure there's more that I don't know of!

**Comment by yuvipanda** [comment]
Can someone with rights mark this as 'good first issue'? :)

**Comment by dhirschfeld** [comment]
I'm interested in RAM - per kernel and total. Not sure if that's already available though?

**Comment by rgbkrk** [comment]
There's some light amount of collaboration around RAM on a kernel level in https://github.com/jupyter/jupyter/issues/264, though it's on a spec level (especially since the actual kernel make be several child processes deep). @ivanov would probably enjoy having a collaborator on the Python side -- I'm looking forward to the UI portion of using it. The notebook server can use it as well since it has access to the messages as it transports them from ZeroMQ to WebSocket.

**Comment by yuvipanda** [comment]
@dhirschfeld @rgbkrk we could possibly also do it from the default Kernel Manager, since it is just spawning local processes and knows how to collect metrics for them (and their children). This lets other kernel managers report their own metrics as they wish, and works across all kernels without any extra work. It would be complimentary to https://github.com/jupyter/jupyter/issues/264.

**Comment by GoelJatin** [comment]
Hi @yuvipanda ,

If no one is working on this, then I would like to take it up.

I would like to start with the very first one for now.

But have a few doubts.

Do we want the number of kernels running at the time when the API is called / keep on collecting them over the complete period of time since the notebook server was started?

Please let me know accordingly.

CC @Madhu94

**Comment by GoelJatin** [comment]
Guys,

Any update on this?

**Comment by manuhortet** [comment]
Hey @GoelJatin, I'm up to work on this with you.

> I'm interested in RAM - per kernel and total. Not sure if that's already available though?

As I think this will be the most useful metric to add, may I take it? :) 

Will try to do it from the default Kernel Manager, as mentioned by @yuvipanda

**Comment by GoelJatin** [comment]
Hey @manuhortet , sure go ahead.

No concerns from my end. :)

**Comment by LiryChen** [comment]
Hey, I am a first-timer looking for tasks to do too. I found this issue pretty interesting. Anything I can help with?

**Comment by konnermacias** [comment]
@manuhortet Have you been able to make any progress on adding that metric? I'm a first-timer as well and would love to help out!

**Comment by manuhortet** [comment]
Hey! Sincerely I've delaying some OS contributions in order to gain time for personal projects. I'm sorry I delayed you two too doing that! You can take this issue if you want to. In fact, feel free to ask me if you face any problems. Good luck! @konnermacias @LiryChen

**Comment by LiryChen** [comment]
Alright thanks! I see look into the issue and may ask you a few questions to understand the problem!

**Comment by Hyaxia** [comment]
@manuhortet Hey, I would be glad to try and help too.
Not sure how all of this is done, but can I just choose one of the options above and start working on it?
If yes then did the guys above decided what they want to work on? So we wont work on the same stuff.

**Comment by manuhortet** [comment]
@Hyaxia of course, you can. Choose some metric you feel relevant from the first comment on this issue and go for it. 
I guess there are already people working on the "_RAM - per kernel and total_" one, so I'd try to avoid that one!

**Comment by Hyaxia** [comment]
@manuhortet Hey, do you have a few minutes spare to help me understand few things?
How will it be best for us to communicate?

**Comment by Hyaxia** [comment]
@manuhortet As far as I can see the "number of running terminals" is not taken, am I right?
If yes, what does that mean a "terminal"?  How does it look in the webapp?

**Comment by Hyaxia** [comment]
#4036

Finished working on the new metrics for the total terminals running currently

**Comment by LiryChen** [comment]
Sorry, I don't think I will continue to work on this due to the limited time I have besides school :( I would like to pick something up in the future once I have more free time!

**Comment by Hyaxia** [comment]
Few questions.
First, I wanted to clarify something about number 4.
Does the kind of tracking that is talked about is tracking the very last thing that the user did and its timestamp?

Second, is anyone still working on the RAM per kernel?

Third, what does number 6 mean?

Thanks.

**Comment by manuhortet** [comment]
For number 4, the last done thing and timestamp would be the logical thing IMO.
Don't think there's anyone working on the RAM per kernel metric, maybe @konnermacias ?

Can't really help on the explanation for number 6. Some help here @yuvipanda ?

**Comment by Hyaxia** [comment]
Ok then, I will start working on the ram per kernel metric in a few days.

**Comment by vinaycalastry** [comment]
Hello.. This issue seems to be open and there has been no status change since October. Can we get the current status on this please ?

**Comment by santosh2702** [comment]
anything i can help with

**Comment by Hyaxia** [comment]
So I guess I'm the last one who was working on it.
When I opened the PR https://github.com/jupyter/notebook/pull/4075, @minrk pointed out that it should go into the jupyter_client project.
Then I opened a PR in jupyter_client https://github.com/jupyter/jupyter_client/pull/407 and you can read yourselves to the point we stopped.

TL;DR - goto https://github.com/jupyter/jupyter_client/pull/407 ,  there should be some kind of generic way to expose different kernel statistics.
I started to work on it in the PR, but just didn't have the time to finish it.
As the last comments there state, check how to do it using `entry points`.
For further information just read the comments in the PR itself, all of the information I had at that time is there.

GL.

**Comment by Franky12** [comment]
Hi can I try this?  looking for some beginner-friendly issues

**Comment by sudo-k-runner** [comment]
Hey! not sure if anyone is still looking into this? I would like to work on this.
